images/openwisp_base/Dockerfile

```diff
@@ -8,93 +8,86 @@ FROM python:3.10-slim-bullseye AS system
     fi
-# TODO: Remove when next version of openwisp-controller is released
-ARG OPENWISP_CONTROLLER_SOURCE=https://github.com/openwisp/openwisp-controller/tarball/3383bc11e5f8e9c89ebd9ac8aefa0066576d3133
 # hadolint ignore=DL3013
 RUN if [ "$OPENWISP_CONTROLLER_SOURCE" != "default" ] ; then \
-        pip install --no-cache-dir --user --upgrade ${OPENWISP_CONTROLLER_SOURCE}; \
+        pip install --no-cache-dir --upgrade ${OPENWISP_CONTROLLER_SOURCE}; \
     fi
-ARG OPENWISP_NOTIFICATION_SOURCE=default
 # hadolint ignore=DL3013
 RUN if [ "$OPENWISP_NOTIFICATION_SOURCE" != "default" ] ; then \
-        pip install --no-cache-dir --user --upgrade ${OPENWISP_NOTIFICATION_SOURCE}; \
+        pip install --no-cache-dir --upgrade ${OPENWISP_NOTIFICATION_SOURCE}; \
     fi
-ARG OPENWISP_USERS_SOURCE=https://github.com/openwisp/openwisp-users/tarball/55023698cf39711585480be90c98d64ce5a04b8d
 # hadolint ignore=DL3013
 RUN if [ "$OPENWISP_USERS_SOURCE" != "default" ] ; then \
-        pip install --no-cache-dir --user --upgrade --force-reinstall ${OPENWISP_USERS_SOURCE}; \
-    fi
-# TODO: Remove when next version of openwisp-utils is released
-ARG OPENWISP_UTILS_SOURCE="openwisp-utils[celery,rest] @ https://github.com/openwisp/openwisp-utils/tarball/9fd347fb861af674674a4bcc59c54bcdeef18b8c "
-# hadolint ignore=DL3013
-RUN if [ "$OPENWISP_UTILS_SOURCE" != "default" ] ; then \
-        pip install --no-cache-dir --user --upgrade --force-reinstall "${OPENWISP_UTILS_SOURCE}"; \
+        pip install --no-cache-dir --upgrade --force-reinstall ${OPENWISP_USERS_SOURCE}; \
     fi
-ARG DJANGO_X509_SOURCE=default
 # hadolint ignore=DL3013
 RUN if [ "$DJANGO_X509_SOURCE" != "default" ]; then \
-        pip install --no-cache-dir --user --upgrade --force-reinstall ${DJANGO_X509_SOURCE}; \
+        pip install --no-cache-dir --upgrade --force-reinstall ${DJANGO_X509_SOURCE}; \
     fi
 
-ARG DJANGO_SOURCE=django~=4.0.10
-# hadolint ignore=DL3013
-RUN pip install --no-cache-dir --user --upgrade ${DJANGO_SOURCE}
-
-COPY ./openwisp_base/requirements.txt /tmp/openwisp-deploy-requirements.txt
-RUN pip install --no-cache-dir --user --upgrade -r /tmp/openwisp-deploy-requirements.txt
-
-FROM system
+FROM system as openwisp_base
 
-COPY --from=openwisp_python --chown=openwisp:root /home/openwisp/.local/ /usr/local
-COPY --chown=openwisp:root ./common/ /opt/openwisp/
-RUN mkdir /opt/openwisp/static && mkdir /opt/openwisp/media && mkdir /opt/openwisp/private && \
-    mkdir /opt/openwisp/logs && chown -R openwisp:root /opt/openwisp
+COPY --from=openwisp_python /usr/local/lib/python3.10/site-packages /usr/local/lib/python3.10/site-packages
+COPY --from=openwisp_python /usr/local/bin /usr/local/bin
+COPY ./common/ /opt/openwisp/
+RUN mkdir /opt/openwisp/static && mkdir /opt/openwisp/media && mkdir /opt/openwisp/private && mkdir /opt/openwisp/logs && \
+    chmod a+w /opt/openwisp /opt/openwisp/static /opt/openwisp/media /opt/openwisp/private /opt/openwisp/logs
```

Comment:
Answers my permission question I guess; I wonder if `a+r` would be enough or not.

Reply:
Unfortunately no because of `collectstatics` that is run at the start and might write to `/opt/openwisp/static` and optionally, the webserver might write to ``/opt/openwisp/media` folder. uWSGIs write to `/opt/openwisp/log` which could be changed to `/var/log/openwisp` wit correct persmissions but I didn't want to do too many changes. And finally `/opt/openwisp/` itself is used to render `uwsgi.conf` during runtime. This could be modified by using `init_command.sh` as the `ENTRYPOINT` and pass parameters to uwsgi via `CMD` array. But again, this would result in even more changes.